### PR TITLE
QL: Add a visitor for traversing recursive evaluations

### PR DESCRIPTION
**Pull request by MathiasVP**
The first commit adds a parameterized module for "folding" over an evaluation of a recursive predicate, and the next commit adds a very simple query that test that the fold operation works as expected.

I think there are likely still some corner cases in the visitor that I haven't got right, but at least it seems to do the right thing most of the time 🤞.

### CODE REVIEW

File: ql/ql/src/codeql_ql/StructuredLogs.qll [hunk 1]
```diff
@@ -317,12 +321,137 @@ module KindPredicatesLog {
     Predicate getPredicate() { result = getPredicateFromPosition(this.getPosition()) }
   }
 
+  /** Gets the `index`'th event that's evaluated by `recursive`. */
+  private InLayer layerEventRank(ComputeRecursive recursive, int index) {
+    result =
+      rank[index + 1](InLayer cand, int startline, string filepath |
+        cand.getComputeRecursiveEvent() = recursive and
+        cand.hasLocationInfo(filepath, startline, _, _, _)
```
Comment (tausbn) [inline ql/ql/src/codeql_ql/StructuredLogs.qll:329]: Hmm... Is there no other way of recovering the ordering? Longer term, I was kinda hoping we could do away with location info for the logs (as the locations relation in the DB is _quite_ big once you include one of the structured logs).

File: ql/ql/src/codeql_ql/StructuredLogs.qll [hunk 2]
```diff
@@ -317,12 +321,137 @@ module KindPredicatesLog {
...
+  }
+
+  /**
+   * Gets the last predicate that's evaluated in an iteration
+   * of the SCC computation rooted at `recursive`.
+   */
+  private InLayer lastPredicate(ComputeRecursive recursive) {
+    exists(int n |
+      result = layerEventRank(recursive, n) and
+      not exists(layerEventRank(recursive, n + 1))
```
Comment (tausbn) [inline ql/ql/src/codeql_ql/StructuredLogs.qll:350]: Hmm... Would it be possible/better to express this using `max`? There's something tickling at the back of my mind that I can't quite articulate.

**Review by tausbn — APPROVED** [review]
So cool! 🚀

File: ql/ql/src/codeql_ql/StructuredLogs.qll (hunk 2 above)
Comment (MathiasVP) [reply ql/ql/src/codeql_ql/StructuredLogs.qll:350]: We could do it with a max as well, yeah. In my experience this join + antijoin pattern patterns better than an aggregation, though.

Comment (MathiasVP) [reply ql/ql/src/codeql_ql/StructuredLogs.qll:350]: But to be fair I didn't actually test it in this case 😂.

Comment (tausbn) [reply ql/ql/src/codeql_ql/StructuredLogs.qll:350]: Interesting. I'll trust your experience, then. 🙂 

(My concern was based on [this](https://github.com/github/codeql/commit/22aa4c93799af788eb27feeae6c690926fafaa3c) bad join in the Python regex libraries, though I guess that's a somewhat different situation.)

Comment (MathiasVP) [reply ql/ql/src/codeql_ql/StructuredLogs.qll:350]: Ah, I see. Yeah, I guess the bad join which that commit is because the `<` and `>` filtering appear too late in the pipeline (which hopefully shouldn't be relevant here).

File: ql/ql/src/codeql_ql/StructuredLogs.qll (hunk 1 above)
Comment (MathiasVP) [reply ql/ql/src/codeql_ql/StructuredLogs.qll:329]: I'm not aware of any other way, unfortunately.

**Comment by MathiasVP** [comment]
Oh, sorry @tausbn. I didn't see that you fixed my PR already 😂. I force pushed a rebase from `main` as well!

**Comment by tausbn** [comment]
No worries. 🙂

**Review by tausbn — APPROVED** [review]
